src/main/java/org/commcare/core/network/AuthenticationInterceptor.java

```diff
@@ -0,0 +1,43 @@
+package org.commcare.core.network;
+
+import java.io.IOException;
+
+import javax.annotation.Nullable;
+
+import okhttp3.Interceptor;
+import okhttp3.Request;
+import okhttp3.Response;
+
+public class AuthenticationInterceptor implements Interceptor {
+
+    @Nullable
+    private String credential;
+
+    @Override
+    public Response intercept(Chain chain) throws IOException {
+        Request original = chain.request();
+
+        // if credentials are null, proceed without authenticating
+        if (credential == null) {
+            return chain.proceed(original);
+        }
+
+        // Throw an exception if we are sending an authenticated request over HTTP
+        if (!original.isHttps() && credential != null) {
```

Comment:
Noting that this would actually be a change to our current behaviors. I think it's likely a good change, but we will need to work to understand if we're introducing limitations that affect meaningful behavior before pushing this.

One example: Internally people do deploys from their local CommCareHQ instance to test functionality, and their machine's don't have HTTPS turned on, so this would prevent that from working

Reply:
So  our earlier `ModernHTTPRequester` was enforcing this. So I retained this in the merger http requestor. By making it controllable do you mean providing some sort of flag in developer options or remove it completely ?

Reply:
Oh, funny! Sorry, I hadn't actually realized we moved over to doing that. If this replicates existing behavior, you can ignore that request.

Reply:
hm, the web team claims the current workflow _is_ working for them (testing with a local machine over http) which I think means that if the old code was trying this that it wasn't working. I'm going to have them test with the new retrofit code and see if it blocks their usage (which would actually be _good_) and then suggest that if that's true we accept authing as long as it is within a local network, IE: to one of

```
localhost
192.168.*
10.*
```
(We can skip the 172 range because it's annoying to encode and no one uses it). 

I actually wonder if any of these utils can just take a URL and answer from some method call whether it's part of a private IP range?

Reply:
So earlier only `ModernHTTPRequester` was enforcing it and not `HTTPRequestGenerator` which was doing majority of requests. `ModernHTTPRequester` was only getting utilized for `PostRequestActivity` and `QueryRequestActivity` which seems to me are rare use cases. This version is definitely going to block HTTP and hence I am going to try to bypass local IPs through this check.

Reply:
thanks. 

I think we may still want a configurable flag for this for non-dimagi domains ("Disable secure authentication requirement" or something), because even though this will be fine for Dimagi domains, any open source project that's using CommCare for something not-secure who don't have HTTPS will essentially have their apps brick after our next release.

Unfortunately I think the way we're going to have to roll this out to not have that problem is a bit complex if we don't want to brick those apps. I think we're going to need to need a feature flag where "Secure Endpoint Enforcement" is disabled by default, and have HQ start adding that flag (as automatically enabled) on all new builds. Alternatively, we can feature flag "Disable Secure Endpoint Enforcement", but set the feature flag to default to _on_ for now, communicate that the change is happening, and then pick a date/release where we set the feature flag to default to "off", which is probably the easiest option.